File: Programmers/Lv3/Lv3_배달.cpp
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

int selectNext(vector<int> cost, vector<bool> check) {
	int end = 0; 
	int min = 500001;
	for (int i = 1; i < cost.size(); i++) {
		if (check[i] == false && cost[i] < min) {		// 비용이 가장 적은 지역을 탐색 
			min = cost[i];
			end = i;
		}
	}
	return end;
}

// 다익스트라 알고리즘 
int solution103(int N, vector<vector<int> > road, int K) {
	int answer = 0;
	vector<map<int, int>> graph(N + 1);	// 비용 정보
	vector<int> cost(N + 1, 500001);		// 비용을 최대치로 초기화
	vector<bool> check(N + 1, false);
	cost[1] = 0;	// 1번은 0으로 초기화

	for (int i = 0; i < road.size(); i++) {		// 1-2, 2-1 두 개로 저장 
		if (graph[road[i][0]].find(road[i][1]) != graph[road[i][0]].end()) {		// 정보가 이미 있다면
			if (graph[road[i][0]][road[i][1]] > road[i][2]) {						// 더 적은 비용을 채택
				graph[road[i][0]][road[i][1]] = road[i][2];
				graph[road[i][1]][road[i][0]] = road[i][2];
			}
		}
		else {		// 정보가 없다면 신규 저장
			graph[road[i][0]].insert(make_pair(road[i][1], road[i][2]));
			graph[road[i][1]].insert(make_pair(road[i][0], road[i][2]));
		}
	}

	for (int i = 0; i < N - 1; i++) {		// N-1만큼 진행 (1번은 0으로 초기화 해뒀음)
		int end = selectNext(cost, check);			// 매번 비용이 가장 적은 지역을 다시 탐색
		for (auto j : graph[end]) {			// 비용이 가장 적은 지역과 연결된 지역들 순회
			if (check[j.first] == false)	// 아직 가보지 않은 곳이라면
				cost[j.first] = min(cost[j.first], cost[end] + j.second);		// 원래 값과 비교하여 저장
		}
		check[end] = true;
	}

	for (int i = 1; i < cost.size(); i++) {
		if (cost[i] <= K)
			answer++;
	}

	return answer;
}

File: Programmers/Lv3/Lv3_배달.py
def search(cost, check):
    end = 0
    temp = 500001
    for i in range(1, len(cost)):
        if check[i] == False and cost[i] < temp:        # 媛???묒? 鍮꾩슜???쒕뒗 留덉쓣 踰덊샇
            temp = cost[i]
            end = i
    return end

def solution(N, road, K):
    answer = 0

    graph = [[500001 for col in range(N+1)] for row in range(N+1)]  # 鍮꾩슜 ?뺣낫
    cost = [500001] * (N + 1)
    check = [False] * (N + 1)
    cost[1] = 0  # 1踰?吏??? 0?쇰줈 珥덇린??

    for i in range(len(road)):
        if graph[road[i][0]][road[i][1]] == 500001:
            graph[road[i][0]][road[i][1]] = road[i][2]
            graph[road[i][1]][road[i][0]] = road[i][2]
        else:
            if graph[road[i][0]][road[i][1]] > road[i][2]:
                graph[road[i][0]][road[i][1]] = road[i][2]
                graph[road[i][1]][road[i][0]] = road[i][2]

    for i in range(1,N+1):
        end = search(cost,check)        # 媛??媛源뚯슫 留덉쓣 李얘린
        for j in range(1, N+1):
            cost[j] = min(cost[j], cost[end] + graph[end][j])
        check[end] = True

    for i in range(1,len(cost)):
        if cost[i] <= K:
            answer += 1
    return answer

print(solution(5,[[1,2,1],[2,3,3],[5,2,2],[1,4,2],[5,3,1],[5,4,2]],3))